File: src/01_data_structures/03_queue/implementations/go/queue.go
package queue

type ArrayQueue[T comparable] struct {
	storage []T
}

func (queue *ArrayQueue[T]) enqueue(value T) *ArrayQueue[T] {
	queue.storage = append(queue.storage, value)

	return queue
}

func (queue *ArrayQueue[T]) dequeue() (T, bool) {
	var zero T

	if len(queue.storage) == 0 {
		return zero, false
	}

	dequeuedElement := queue.storage[0]
	queue.storage = queue.storage[1:]

	return dequeuedElement, true
}

func (queue ArrayQueue[T]) isEmpty() bool {
	return len(queue.storage) == 0
}

func (queue ArrayQueue[T]) length() int {
	return len(queue.storage)
}

func (queue ArrayQueue[T]) peek() T {
	return queue.storage[0]
}

File: src/01_data_structures/03_queue/implementations/ts/queue.ts
export class ArrayQueue<T> {
  private storage: Array<T> = [];

  enqueue(value: any): ArrayQueue<T> {
    this.storage.push(value);

    return this;
  }

  dequeue(): T | null {
    if (this.storage.length === 0) {
      return null;
    }

    return this.storage.shift();
  }

  length(): number {
    return this.storage.length;
  }

  isEmpty(): boolean {
    return this.length() === 0;
  }

  peek(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    return this.storage[0];
  }
}
